### ISSUE TITLE
[POC for Node]Using Deep and Wide Neural Networks to generate stack recommendations

### ISSUE THREAD

**Issue by abs51295** [openshiftio/openshift.io#1803]
#### User Story

As an OSIO developer, I will be able to see stack recommendations via the new approach so that I can improve my stack.

#### Description

Currently, companion recommendations come from PGM using hierarchical Bayesian inference and on the other hand the outlier recommendations come from the usage data of the packages. Using [Deep and Wide networks](https://www.tensorflow.org/tutorials/wide_and_deep) we can leverage the power of both memorization and inference to give better recommendations to the end user.

#### Acceptance Criteria

A working POC with multiple fine tuned neural networks implemented and tested. The results/requirements/shortcomings of each for the NPM ecosystem should be documented at the end of this spike.

##### Using the existing data available for NPM:

- [ ] Implement a deep and wide neural network model using the sample code available in the [tutorial](https://www.tensorflow.org/tutorials/wide_and_deep).

    - [ ] Check for over-fitting.
    - [ ] Calculate the precision of the implemented model.
    - [ ] Fine tune the hyper parameters and iterate over the implementation of the model by                          repeating the previous steps.

##### Using the existing data available for Python:

- [ ] Implement a wide neural network model.
  - [ ] Check for over-fitting of the model.
  - [ ] Calculate the precision of the implemented model.
  - [ ] Fine tune the hyper parameters and iterate over the implementation of the model by                          repeating the previous steps.

- [ ] Document the results of both the above approaches.

##### Collection of data for NPM and re-training of the models:

- [x] Collect data for the NPM ecosystem.
- [ ] Re-train the models using the newly collected data.
- [ ] Check for precision of the re-implemented models and iterate if required.
- [ ] Document the findings.

**Comment by abs51295** [comment]
The first part for NPM data collection is done with help from @miteshvp . Data is now available in s3.

**Comment by abs51295** [comment]
The Deep and Wide model as described in this [research paper](https://arxiv.org/abs/1606.07792), has two parts:

1) Retrieval System:
    - The retrieval system returns a short list of items that best match the query using various signals, usually a combination of machine-learned models and human-defined rules.

2) Ranking System:
    - The ranking system ranks all items received from the retrieval system by their scores. The scores are usually P(y|x), the probability of a user action label y given the features x, including user features (e.g., country, language, demographics), contextual features (e.g., device, hour of the day, day of the week), and impression features (e.g., app age, historical statistics of an app).

The paper describes how to model the ranking system while assuming that we have the retrieval system in place which is not the case. We can loop back to this once we have a good retrieval system in place. Closing this for now. Thanks.
